source/common/ssl/context_impl.cc

```diff
@@ -84,19 +103,52 @@ ContextImpl::ContextImpl(ContextManagerImpl& parent, Stats::Scope& scope,
     SSL_CTX_set_cert_verify_callback(ctx_.get(), ContextImpl::verifyCallback, this);
   }
 
-  if (!config.certChainFile().empty()) {
-    cert_chain_ = loadCert(config.certChainFile());
-    cert_chain_file_path_ = config.certChainFile();
-    int rc = SSL_CTX_use_certificate_chain_file(ctx_.get(), config.certChainFile().c_str());
-    if (0 == rc) {
+  if (config.certChain().empty() != config.privateKey().empty()) {
+    throw EnvoyException(fmt::format("Failed to load incomplete certificate from {}, {}",
+                                     config.certChainPath(), config.privateKeyPath()));
+  }
+
+  if (!config.certChain().empty()) {
+    // Load certificate chain.
+    cert_chain_file_path_ = config.certChainPath();
+    bssl::UniquePtr<BIO> bio(
+        BIO_new_mem_buf(const_cast<char*>(config.certChain().data()), config.certChain().size()));
+    RELEASE_ASSERT(bio != nullptr);
+    cert_chain_.reset(PEM_read_bio_X509_AUX(bio.get(), nullptr, nullptr, nullptr));
+    if (cert_chain_ == nullptr || !SSL_CTX_use_certificate(ctx_.get(), cert_chain_.get())) {
+      throw EnvoyException(
+          fmt::format("Failed to load certificate chain from {}", config.certChainPath()));
+    }
+    // Read rest of the certificate chain.
+    for (;;) {
```

Comment:
"while (true)" is the convention used in this codebase

Reply:
Done.